### PR TITLE
fix: Long path handling for CREATE_COPY file conflict resolution during job download

### LINKED ISSUE #617: Bug: Long path handling doesn't work for output copies that are downloaded
### Describe the bug

If an output is downloaded twice using ``download-output`` cli with ``--conflict-resolution`` as ``CREATE_COPY`` a copy is created such as ``image (1).png``. However, if such a file becomes a long path for Windows after this unique name is chosen, it is not handled currently.

### Expected Behaviour

Long paths should be handled for these cases after a unique name is chosen, it is being done before right now which may result in edge cases where the new unique name is a long path but the original name wasn't

### Current Behaviour

If the unique copy's name with the number appended to it is a long path, it is not handled currently.

### Reproduction Steps

Run ``deadline job download-output --job-id job-xx --farm-id farm-xx --queue-id queue-xx --conflict-resolution CREATE_COPY
`` on Windows such that the path becomes a long path for an output. Pick a long file name more than allowed character on windows

### Environment

At minimum:
1. Operating system (e.g. Windows Server 2022; Amazon Linux 2023; etc.) Windows
2. Output of `python3 --version` does not matter
3. Output of `deadline --version` 0.49.6.
4. If this is from a version installed by the Deadline Cloud Submitter installer, then what version of the submitter installer? N/A

Please share other details about your environment that you think might be relevant to reproducing the bug.

### PR DESCRIPTION
**Pull request by sakshie95**
Fixes: https://github.com/aws-deadline/deadline-cloud/issues/617

### What was the problem/requirement? (What/Why)
If an output is downloaded twice using ``download-output`` cli with ``--conflict-resolution`` as ``CREATE_COPY`` a copy is created such as ``image (1).png``. However, if such a file becomes a long path for Windows after this unique name is chosen, it is not handled currently.

### What was the solution? (How)
Added long path handling for a file after its name is updated with the create copy nomenclature.

### What is the impact of this change?
Handled edge case so that long paths are handled for all cases.

### How was this change tested?
Wrote & ran unit tests for windows and posix.
Windows with UNC path handling:
```
=================================================================================== slowest 5 durations ====================================================================================
1.74s setup    test/unit/deadline_job_attachments/test_download.py::TestFullDownload::test_download_file_create_copy_becomes_long_path_windows[2023-03-03]
(2 durations < 0.005s hidden.  Use -vv to show these durations.)
============================================================================== 1 passed, 2 warnings in 9.75s ===============================================================================
P
```
Posix:
```
[gw0] [100%] PASSED test/unit/deadline_job_attachments/test_download.py::TestFullDownload::test_download_file_create_copy_long_path_posix[2023-03-03] 

============================= slowest 5 durations ==============================
0.74s setup    test/unit/deadline_job_attachments/test_download.py::TestFullDownload::test_download_file_create_copy_long_path_posix[2023-03-03]

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
============================== 1 passed in 3.64s ===============================
```

See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#testing) for information on running tests.

- Have you run the unit tests? yes
- Have you run the integration tests? yes
- Have you made changes to the `download` or `asset_sync` modules? If so, then it is highly recommended
  that you ensure that the docker-based unit tests pass. N/A

### Was this change documented?

- Are relevant docstrings in the code base updated? N/A
- Has the README.md been updated? If you modified CLI arguments, for instance. N/A

### Does this PR introduce new dependencies?

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies).
*   [x] This PR does not add any new dependencies.

### Is this a breaking change?
No

### Does this change impact security?
No
----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/job_attachments/download.py [hunk 1]
```diff
@@ -15,65 +15,65 @@
 from logging import Logger, LoggerAdapter, getLogger
 from pathlib import Path
 from tempfile import NamedTemporaryFile
-from typing import Any, Callable, DefaultDict, List, Optional, Tuple, Union
+from threading import Lock
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py]: Why the formatting change? the default formatter should have caught everything on the previous submit.

File: src/deadline/job_attachments/download.py [hunk 2]
```diff
@@ -550,9 +551,13 @@ def download_file(
...
             pass
         elif file_conflict_resolution == FileConflictResolution.CREATE_COPY:
-            local_file_path = _get_new_copy_file_path(
+            copy_local_file_path = _get_new_copy_file_path(
                 local_file_path, collision_lock, collision_file_dict
             )
+
+            # Re-run _get_long_path_compatible_path for updated file name after file conflict resolution
+            # _get_long_path_compatible_path is idempotent, so it doesn't re-process an existing long path
+            local_file_path = _get_long_path_compatible_path(copy_local_file_path)
```
Comment (leongdl) [inline src/deadline/job_attachments/download.py:559]: Next time put a comment here with the link so I can see what this does. https://github.com/aws-deadline/deadline-cloud/blob/7e51bb574bae31ba6bbeea579306a81184aab762/src/deadline/job_attachments/_utils.py#L111

File: src/deadline/job_attachments/download.py (hunk 1 above)
Comment (sakshie95) [reply src/deadline/job_attachments/download.py]: I agree but something was up with ``hatch fmt`` which got me into this formatting pit. I'm trying something to see if I can resolve this differently

File: src/deadline/job_attachments/download.py (hunk 2 above)
Comment (sakshie95) [reply src/deadline/job_attachments/download.py:559]: cool, https://github.com/aws-deadline/deadline-cloud/blob/7e51bb574bae31ba6bbeea579306a81184aab762/src/deadline/job_attachments/_utils.py#L111 :))

File: test/unit/deadline_job_attachments/test_download.py [hunk 3]
```diff
@@ -2059,6 +2059,154 @@ def test_windows_long_path_UNC_notation_and_registry_WindowsOS(self):
...
+        Tests that original path is not long but copy becomes long.
+        """
+        original_file = Path(base_dir) / f"{long_base_name}.txt"
+        copy_file = Path(base_dir) / f"{long_base_name} (1).txt"
+
+        # Verify our test scenario is correct
+        original_len = len(str(original_file)) + 9
+        copy_len = len(str(copy_file)) + 9
+        assert original_len < 260, f"Original should NOT be long path: {original_len}"
+        assert copy_len >= 260, f"Copy should become long path: {copy_len}"
```
Comment (jericht) [inline test/unit/deadline_job_attachments/test_download.py]: nit: where does the 260 come from?

File: test/unit/deadline_job_attachments/test_download.py [hunk 4]
```diff
@@ -2059,6 +2059,154 @@ def test_windows_long_path_UNC_notation_and_registry_WindowsOS(self):
...
+
+        # Create the original file to force a conflict
+        original_file.parent.mkdir(parents=True, exist_ok=True)
+        original_file.write_text("original content")
```
Comment (jericht) [inline test/unit/deadline_job_attachments/test_download.py]: Do we need this? We're already simulating with the mock `is_file()` below.

File: test/unit/deadline_job_attachments/test_download.py [hunk 5]
```diff
@@ -2059,6 +2059,154 @@ def test_windows_long_path_UNC_notation_and_registry_WindowsOS(self):
...
+        self._test_create_copy_long_path_scenario(base_dir, long_base_name, expect_unc_prefix=True)
+
+    @pytest.mark.skipif(
+        sys.platform == "win32",
+        reason="This test is for POSIX systems.",
+    )
+    @pytest.mark.parametrize(
+        "dir_multiplier,filename_len",
+        [
+            (14, 85) if sys.platform == "darwin" else (22, 72),
```
Comment (godobyte) [inline test/unit/deadline_job_attachments/test_download.py:2195]: nit - this could use some comment to explain.

**Review by godobyte — DISMISSED** [review]

File: test/unit/deadline_job_attachments/test_download.py (hunk 3 above)
Comment (sakshie95) [reply test/unit/deadline_job_attachments/test_download.py]: It's the Windows Max path length. See https://github.com/aws-deadline/deadline-cloud/blob/7e51bb574bae31ba6bbeea579306a81184aab762/src/deadline/job_attachments/_utils.py#L29. Maybe I can use this constant to make it clearer

Comment (sakshie95) [reply test/unit/deadline_job_attachments/test_download.py]: Done

File: test/unit/deadline_job_attachments/test_download.py (hunk 5 above)
Comment (sakshie95) [reply test/unit/deadline_job_attachments/test_download.py:2195]: I've added it

File: test/unit/deadline_job_attachments/test_download.py (hunk 4 above)
Comment (sakshie95) [reply test/unit/deadline_job_attachments/test_download.py]: You're right I was not mocking it before but did so later. I've removed this

**Review by godobyte — APPROVED** [review]

**Review by jericht — APPROVED** [review]
